### PR TITLE
[Glimmer2] port remaining if-unless tests

### PR DESCRIPTION
**Pull request by GavinJoyce**
Part of https://github.com/emberjs/ember.js/issues/13127

TODO:
- [x] port, consolidate or delete remaining tests
- [x] annotate/deleted moved tests

### CODE REVIEW

**Comment by rwjblue** [comment]
@GavinJoyce - Any thoughts on this?  I haven't been able to follow along quite as well as I intended, I'm not sure if this is finished or still WIP...

**Comment by rwjblue** [comment]
Travis failed for `packages/ember-htmlbars/tests/helpers/if_unless_test.js` still including the `jQuery` import when it isn't used any more.

**Comment by GavinJoyce** [comment]
@rwjblue this is still a WIP. I'll bring it to completion tomorrow afternoon

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 1]
```diff
@@ -1,260 +0,0 @@
...
-    owner = view = null;
-
-    Ember.lookup = lookup = originalLookup;
-    TemplateTests = null;
-
-    resetKeyword('view', originalViewKeyword);
-  }
-});
-
-QUnit.test('properties within an if statement should not fail on re-render', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: deleted as this is covered by `Syntax test: {{#if}} and {{#unless}} without inverse` tests

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 2]
```diff
@@ -1,260 +0,0 @@
...
-  equal(view.$().text(), 'test');
-
-  run(function() {
-    view.set('value', null);
-  });
-
-  equal(view.$().text(), '');
-});
-
-QUnit.test('views within an if statement should be sane on re-render', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: deleted as this is testing old views, please let me know if you think that this is incorrect

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 3]
```diff
@@ -1,260 +0,0 @@
...
-  });
-
-  var textfield = view.$('input');
-  equal(textfield.length, 1);
-
-  // Make sure the view is still registered in View.views
-  ok(EmberView.views[textfield.attr('id')]);
-});
-
-QUnit.test('the {{this}} helper should not fail on removal', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: deleted, please let me know if you think that this is incorrect

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 4]
```diff
@@ -1,260 +0,0 @@
...
-  equal(view.$().text(), 'abc', 'should start property - precond');
-
-  run(function() {
-    view.set('show', false);
-  });
-
-  equal(view.$().text(), '');
-});
-
-QUnit.test('properties within an if statement should not fail on re-render', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: deleted as this is covered by `Syntax test: {{#if}} and {{#unless}} without inverse` tests

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 5]
```diff
@@ -1,260 +0,0 @@
...
-  equal(view.$().text(), 'test');
-
-  run(function() {
-    view.set('value', null);
-  });
-
-  equal(view.$().text(), '');
-});
-
-QUnit.test('the {{this}} helper should not fail on removal', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: this test was duplicated, see `the {{this}} helper should not fail on removal` above

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 6]
```diff
@@ -1,260 +0,0 @@
...
-  equal(view.$().text(), 'abc', 'should start property - precond');
-
-  run(function() {
-    view.set('show', false);
-  });
-
-  equal(view.$().text(), '');
-});
-
-QUnit.test('edge case: child conditional should not render children if parent conditional becomes false', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: moved to `@test child conditional should not render children if parent conditional becomes false`

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 7]
```diff
@@ -1,260 +0,0 @@
...
-    view.set('cond1', false);
-  });
-
-  // TODO: Priority Queue, for now ensure correct result.
-  ok(!childCreated, 'child should not be created');
-  //ok(child.isDestroyed, 'child should be gone');
-  equal(view.$().text(), '');
-});
-
-QUnit.test('edge case: rerender appearance of inner virtual view', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: deleted as this is covered by `Syntax test: {{#if}} and {{#unless}} without inverse` tests

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 8]
```diff
@@ -1,260 +0,0 @@
...
-  equal(jQuery('#qunit-fixture').text(), '');
-
-  run(function() {
-    view.set('cond2', true);
-  });
-
-  equal(jQuery('#qunit-fixture').text(), 'test');
-});
-
-QUnit.test('`if` helper with inline form: can use truthy param as binding', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: deleted, covered by `Helpers test: inline {{if}}`

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js [hunk 9]
```diff
@@ -1,260 +0,0 @@
...
-  equal(view.$().text(), 'ok');
-
-  run(function() {
-    view.set('truthy', 'yes');
-  });
-
-  equal(view.$().text(), 'yes');
-});
-
-QUnit.test('using `if` with an `{{each}}` destroys components when transitioning to and from inverse (GH #12267)', function() {
```
Comment (GavinJoyce) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: ported to `@test using `if` with an `{{each}}` destroys components when transitioning to and from inverse (GH #12267)`

**Comment by GavinJoyce** [comment]
@chancancode this is ready for review

**Comment by GavinJoyce** [comment]
TODO:

 * [ ] rebase

**Comment by GavinJoyce** [comment]
rebased

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 1 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :heavy_check_mark:

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 2 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :heavy_check_mark:

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 3 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :heavy_check_mark:

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 4 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :heavy_check_mark:

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 5 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :heavy_check_mark:

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 6 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :+1: :+1: :+1:

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 7 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :heavy_check_mark:

File: packages/ember-glimmer/tests/integration/syntax/if-unless-test.js [hunk 10]
```diff
@@ -24,3 +28,85 @@ moduleFor('Syntax test: {{#if}} and {{#unless}} without inverse', class extends
   }
 
 });
+
+moduleFor('Syntax test: {{#if}}', class extends RenderingTest {
+
+  ['@test child conditional should not render children if parent conditional becomes false'](assert) {
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/syntax/if-unless-test.js]: I suspect this could be hoisted up to the `TogglingSyntaxConditionalsTest` level (maybe even the `TogglingConditionalsTest` if we make `foo-bar` a helper instead). Maybe something like this would work:

```javascript
let innerTemplate = this.templateFor({ cond: 'cond2', truthy: '{{foo-bar}}', falsy: '' }); // {{#if cond2}}{{foo-bar}}{{else}}{{/if}}
let outerTemplate = this.templateFor({ cond: 'cond1', truthy: innerTemplate, falsy: '' }); // {{#if cond1}}{{#if cond2}}{{foo-bar}}{{else}}{{/if}}{{else}}{{/if}}
let wrappedTemplate = this.wrapperFor([outerTemplate]);

this.render(wrappedTemplate, { cond1: true, cond2: false });

...

this.runTask(() => {
  set(this.context, 'cond2', true);
  set(this.context, 'cond1', false);
});

...
```

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 8 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :heavy_check_mark:

File: packages/ember-glimmer/tests/integration/syntax/if-unless-test.js [hunk 11]
```diff
@@ -1,4 +1,8 @@
-import { moduleFor } from '../../utils/test-case';
+import Component from 'ember-views/components/component';
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/syntax/if-unless-test.js]: `import { Component } from '../../utils/helpers';`

File: packages/ember-glimmer/tests/integration/syntax/if-unless-test.js [hunk 12]
```diff
@@ -24,3 +28,85 @@ moduleFor('Syntax test: {{#if}} and {{#unless}} without inverse', class extends
...
+      template: '{{number}}',
+      ComponentClass: Component.extend({
+        willDestroy() {
+          this._super();
+          destroyedChildrenCount++;
+        }
+      })
+    });
+
+    this.render(`
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/syntax/if-unless-test.js]: We can use the `strip` helper instead of using `{{~...~}}`, see https://github.com/emberjs/ember.js/blob/master/packages/ember-glimmer/tests/integration/application/rendering-test.js#L23-L29

File: packages/ember-htmlbars/tests/helpers/if_unless_test.js (hunk 9 above)
Comment (chancancode) [inline packages/ember-htmlbars/tests/helpers/if_unless_test.js]: :heavy_check_mark:

**Comment by chancancode** [comment]
Looks great to me (as usual)! We need to fix the `Component` import before merging, the other comments are just stretch goals

File: packages/ember-glimmer/tests/integration/syntax/if-unless-test.js (hunk 10 above)
Comment (GavinJoyce) [inline packages/ember-glimmer/tests/integration/syntax/if-unless-test.js]: Nice idea, I've moved it to `TogglingSyntaxConditionalsTest`. I didn't manage to get it to work in `TogglingConditionalsTest`, I've added this to my list of things to revisit soon though

File: packages/ember-glimmer/tests/integration/syntax/if-unless-test.js (hunk 11 above)
Comment (GavinJoyce) [inline packages/ember-glimmer/tests/integration/syntax/if-unless-test.js]: done

File: packages/ember-glimmer/tests/integration/syntax/if-unless-test.js (hunk 12 above)
Comment (GavinJoyce) [inline packages/ember-glimmer/tests/integration/syntax/if-unless-test.js]: TIL, thanks. Done
